### PR TITLE
[mlir][tensor] fix out-of-bound index in tensor.dim

### LINKED ISSUE #70183: [mlir][tensor] --canonicalize crashed with assertion failure "invalid index for shaped type"
git version: b745ce952560716545e82770bf835e48a567eaf1

system: Ubuntu 20.04.6 LTS (Focal Fossa)

reproduced with: mlir-opt --canonicalize a.mlir

a.mlir:

```
module {
  func.func private @func2() {
    %c1 = arith.constant 1 : index
    %idx28 = index.constant 28
    %c29 = arith.constant 29 : index
    %3 = bufferization.alloc_tensor(%c29) : tensor<?xi16>
    // The crash would disappear, if %idx28 is 0. So the crash may due to UB in tensor.dim.
    %dim = tensor.dim %3, %idx28 : tensor<?xi16>
    %alloc_21 = memref.alloc(%c29) : memref<?x26x2xi32>
    %16 = affine.vector_load %alloc_21[%c1, %c1, %dim] : memref<?x26x2xi32>, vector<7xi32>
    return
  }
}
```

trace:

```
mlir-opt: /data/bin/llvm-project/build/tools/mlir/include/mlir/IR/BuiltinTypeInterfaces.h.inc:240: bool mlir::detail::ShapedTypeTrait<ConcreteType>::isDynamicDim(unsigned int) const [with ConcreteType = mlir::RankedTensorType]: Assertion `idx < getRank() && "invalid index for shaped type"' failed.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /data/bin/llvm-project/build/bin/mlir-opt --canonicalize a.mlir
 #0 0x0000556306ca73af llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/data/bin/llvm-project/build/bin/mlir-opt+0x15ca3af)
 #1 0x0000556306ca4444 SignalHandler(int) Signals.cpp:0:0
 #2 0x00007f5329672420 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x14420)
 #3 0x00007f532915d00b raise (/lib/x86_64-linux-gnu/libc.so.6+0x4300b)
 #4 0x00007f532913c859 abort (/lib/x86_64-linux-gnu/libc.so.6+0x22859)
 #5 0x00007f532913c729 (/lib/x86_64-linux-gnu/libc.so.6+0x22729)
 #6 0x00007f532914dfd6 (/lib/x86_64-linux-gnu/libc.so.6+0x33fd6)
 #7 0x000055630711bb9d (/data/bin/llvm-project/build/bin/mlir-opt+0x1a3eb9d)
 #8 0x000055630711c4e2 (anonymous namespace)::FoldDimOfAllocTensorOp::matchAndRewrite(mlir::tensor::DimOp, mlir::PatternRewriter&) const BufferizationOps.cpp:0:0
 #9 0x00005563071100aa mlir::detail::OpOrInterfaceRewritePatternBase<mlir::tensor::DimOp>::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const (/data/bin/llvm-project/build/bin/mlir-opt+0x1a330aa)
#10 0x000055630c4c20b8 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>) (/data/bin/llvm-project/build/bin/mlir-opt+0x6de50b8)
#11 0x00005563098a216c (anonymous namespace)::GreedyPatternRewriteDriver::processWorklist() GreedyPatternRewriteDriver.cpp:0:0
#12 0x00005563098a65db mlir::applyPatternsAndFoldGreedily(mlir::Region&, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig, bool*) (/data/bin/llvm-project/build/bin/mlir-opt+0x41c95db)
#13 0x000055630980c39a (anonymous namespace)::Canonicalizer::runOnOperation() Canonicalizer.cpp:0:0
#14 0x00005563097ed39e mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) (/data/bin/llvm-project/build/bin/mlir-opt+0x411039e)
#15 0x00005563097ed87a mlir::detail::OpToOpPassAdaptor::runPipeline(mlir::OpPassManager&, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) (/data/bin/llvm-project/build/bin/mlir-opt+0x411087a)
#16 0x00005563097ee0a4 mlir::PassManager::run(mlir::Operation*) (/data/bin/llvm-project/build/bin/mlir-opt+0x41110a4)
#17 0x00005563097de96b performActions(llvm::raw_ostream&, std::shared_ptr<llvm::SourceMgr> const&, mlir::MLIRContext*, mlir::MlirOptMainConfig const&) MlirOptMain.cpp:0:0
#18 0x00005563097dfe24 processBuffer(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::MlirOptMainConfig const&, mlir::DialectRegistry&, llvm::ThreadPool*) MlirOptMain.cpp:0:0
#19 0x00005563097dffd4 mlir::LogicalResult llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>::callback_fn<mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&)::'lambda'(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>(long, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&) MlirOptMain.cpp:0:0
#20 0x00005563098dc8e4 mlir::splitAndProcessBuffer(std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<mlir::LogicalResult (std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, llvm::raw_ostream&)>, llvm::raw_ostream&, bool, bool) (/data/bin/llvm-project/build/bin/mlir-opt+0x41ff8e4)
#21 0x00005563097d92a7 mlir::MlirOptMain(llvm::raw_ostream&, std::unique_ptr<llvm::MemoryBuffer, std::default_delete<llvm::MemoryBuffer>>, mlir::DialectRegistry&, mlir::MlirOptMainConfig const&) (/data/bin/llvm-project/build/bin/mlir-opt+0x40fc2a7)
#22 0x00005563097e02fb mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&) (/data/bin/llvm-project/build/bin/mlir-opt+0x41032fb)
#23 0x0000556306bdfa1b main (/data/bin/llvm-project/build/bin/mlir-opt+0x1502a1b)
#24 0x00007f532913e083 __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x24083)
#25 0x0000556306c77cde _start (/data/bin/llvm-project/build/bin/mlir-opt+0x159acde)
Aborted (core dumped)
```

### PR DESCRIPTION
**Pull request by dynamicheart**
fold tensor.dim with out-of-bound index to ub.poison

Fixes: https://github.com/llvm/llvm-project/issues/70183

### CODE REVIEW

**Comment by dynamicheart** [comment]
@lipracer please also review this PR.

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp [hunk 1]
```diff
@@ -738,11 +741,11 @@ OpFoldResult DimOp::fold(FoldAdaptor adaptor) {
   if (!tensorType)
     return {};
 
-  // Out of bound indices produce undefined behavior but are still valid IR.
-  // Don't choke on them.
+  // Fold dim to posion if the index is out of bound. Poison represents
```
Comment (joker-eph) [inline mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:744]:
```suggestion
  // Fold dim to poison if the index is out of bound. Poison represents
```

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp [hunk 2]
```diff
@@ -738,11 +741,11 @@ OpFoldResult DimOp::fold(FoldAdaptor adaptor) {
...
+  // undefined behavior.
   int64_t indexVal = index.getInt();
   if (indexVal < 0 || indexVal >= tensorType.getRank())
-    return {};
+    return ub::PoisonAttr::get(getContext());
```
Comment (joker-eph) [inline mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:748]: Is the Tensor dialect already dependent on UB dialect and generating poison values?

**Comment by lipracer** [comment]
You need to add a dependency on UB dialect in cmake file like [this](https://github.com/llvm/llvm-project/blob/6960ace534c4021301dd5a9933ca06ba96edea23/mlir/lib/Dialect/Affine/IR/CMakeLists.txt#L23).

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp [hunk 3]
```diff
@@ -738,11 +741,11 @@ OpFoldResult DimOp::fold(FoldAdaptor adaptor) {
   if (!tensorType)
     return {};
 
-  // Out of bound indices produce undefined behavior but are still valid IR.
-  // Don't choke on them.
+  // Fold dim to posion if the index is out of bound. Poison represents
+  // undefined behavior.
```
Comment (lipracer) [inline mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:745]: // Undefined behavior.

**Review by lipracer — APPROVED** [review]
LGTM.If the suggestions have been fixed.

File: mlir/lib/Dialect/Tensor/IR/TensorOps.cpp (hunk 2 above)
Comment (joker-eph) [reply mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:748]: This is really a design question here: this new addition can be a problem for some user, it's not clear to me if we can actually use poison everywhere in MLIR right now.

Comment (dynamicheart) [reply mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:748]: I read the code comment again and think you are right.  It seems `tensor.dim` ops with out-of-bound indices are treated as valid IR and may have following handling solutions:
- Compiled to runnable binary and raise exception on runtime
- Handled in other passes to raise warning or errors on compile time depends on the compiler design.

So should I just simply fix the out-of-bound access bug in the PatternRewriter `FoldDimOfAllocTensorOp` and keep this folding behavior as before. What's your opinion?

Comment (dynamicheart) [reply mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:748]: Another question: is `ub.poison` is designed to be a valid IR and able to be compiled to a binary?

Comment (lipracer) [reply mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:748]: After reviewing the design related to Poison semantics, there is no clear definition of the issue of excessive crossing. Can we simply skip folding and do nothing according to the original logic, so as not to disrupt the processing of the original code.

Comment (joker-eph) [reply mlir/lib/Dialect/Tensor/IR/TensorOps.cpp:748]: >  It seems tensor.dim ops with out-of-bound indices are treated as valid IR 

It is documented as undefined behavior, so turning it into poison is clearly allowed by the spec.
However at the moment we haven't deployed the UB dialect as a really "central" component, and "fold" is, even more than canonicalization, very anchored in the core of MLIR (it's not optional, can't be disabled, etc.). 
So folding to poison makes me a bit nervous here, if there were already other folding of the tensor dialect to poison that wouldn't be a problem.

> Can we simply skip folding and do nothing 

This is the most conservative solution IMO.

**Comment by lipracer** [comment]
@dynamicheart Thanks for fixsed,  Please wait ci green,I will merge this.

**Pull request merged**